File: datagen/src/main/java/juuxel/adorn/datagen/Id.java
package juuxel.adorn.datagen;

import java.util.regex.Pattern;

/**
 * A namespaced ID. Mimics {@code net.minecraft.resources.Identifier}.
 */
public record Id(String namespace, String path) {
    private static final Pattern NAMESPACE_PATTERN = Pattern.compile("^[a-z0-9._-]+$");
    private static final Pattern PATH_PATTERN = Pattern.compile("^[a-z0-9._/-]+$");

    public Id {
        if (!NAMESPACE_PATTERN.matcher(namespace).matches()) {
            throw new IllegalArgumentException(
                "ID namespace '%s' doesn't match %s".formatted(namespace, NAMESPACE_PATTERN)
            );
        } else if (!PATH_PATTERN.matcher(path).matches()) {
            throw new IllegalArgumentException(
                "ID path '%s' doesn't match %s".formatted(path, PATH_PATTERN)
            );
        }
    }

    public Id rawSuffixed(String suffix) {
        return new Id(namespace, path + suffix);
    }

    public Id suffixed(String suffix) {
        return rawSuffixed("_" + suffix);
    }

    public Id rawPrefixed(String prefix) {
        return new Id(namespace, prefix + path);
    }

    public Id prefixed(String prefix) {
        return rawPrefixed(prefix + "_");
    }

    @Override
    public String toString() {
        return namespace + ':' + path;
    }

    public static Id parse(String id) {
        int idIndex = id.indexOf(':');
        if (idIndex < 0) {
            throw new IllegalArgumentException("Id '%s' does not contain colon (:)!".formatted(id));
        }
        var ns = id.substring(0, idIndex);
        var path = id.substring(idIndex + 1);
        return new Id(ns, path);
    }
}

File: datagen/src/main/kotlin/juuxel/adorn/datagen/Id.kt
package juuxel.adorn.datagen

import java.io.Serializable

/**
 * A namespaced ID. Mimics `net.minecraft.util.Identifier`.
 */
data class Id(val namespace: String, val path: String) : Serializable {
    init {
        require(validateIdCharacters(namespace, isPath = false)) {
            "ID namespace '$namespace' contains characters not matching [a-z0-9.-_]"
        }
        require(validateIdCharacters(path, isPath = true)) {
            "ID path '$path' contains characters not matching [a-z0-9.-_/]"
        }
    }

    /**
     * Suffixes the path with a [suffix] without separation.
     */
    fun rawSuffixed(suffix: String): Id = Id(namespace, path + suffix)

    /**
     * Suffixes the path with an underscore-separated [suffix].
     */
    fun suffixed(suffix: String): Id = rawSuffixed("_$suffix")

    override fun toString() = "$namespace:$path"

    companion object {
        fun parse(id: String): Id {
            require(':' in id) { "Id '$id' does not contain colon (:)!" }
            val parts = id.split(":")
            require(parts.size == 2) { "Id '$id' must consist of exactly two parts! Found: ${parts.size}" }
            val (ns, path) = parts
            return Id(ns, path)
        }

        private fun validateIdCharacters(component: String, isPath: Boolean): Boolean =
            component.all {
                it in 'a'..'z' || it in '0'..'9' || it == '.' || it == '-' || it == '_' || (isPath && it == '/')
            }
    }
}
